lib/proxy/peer/credentials.go

```diff
@@ -0,0 +1,207 @@
+	if err := validatePeer(c.peerID, identity); err != nil {
+		c.logger.Errorf(duplicatePeerMsg, c.peerAddr, c.peerID)
+		return nil, nil, trace.Wrap(err)
+	}
+
+	return conn, authInfo, nil
+}
+
+// getIdentity returns a [tlsca.Identity] that is created from the certificate
+// presented during the TLS handshake.
+func getIdentity(authInfo credentials.AuthInfo) (*tlsca.Identity, error) {
+	tlsInfo, ok := authInfo.(credentials.TLSInfo)
+	if !ok {
+		return nil, trace.AccessDenied("credentials auth information is missing")
+	}
+
+	certs := tlsInfo.State.PeerCertificates
+	if len(certs) == 0 {
+		return nil, trace.AccessDenied("no peer certificates provided")
+	}
+
+	clientCert := certs[0]
+	identity, err := tlsca.FromSubject(clientCert.Subject, clientCert.NotAfter)
+	return identity, trace.Wrap(err)
+}
+
+// checkProxyRole ensures that the [tlsca.identity] is for a [types.RoleProxy].
+func checkProxyRole(identity *tlsca.Identity) error {
+	for _, role := range identity.Groups {
+		if types.SystemRole(role) == types.RoleProxy {
+			return nil
+		}
+	}
+
+	return trace.AccessDenied("proxy system role required")
+}
+
+// validatePeer ensures that provided peerID matches the id of
+// the peer that was connected to. This prevents client connections
+// from being established to an incorrect peer if multiple peers
+// share the same address.
+func validatePeer(peerID string, identity *tlsca.Identity) error {
+	if identity.Username == peerID {
+		return nil
+	}
+
+	return trace.AccessDenied("connected to unexpected proxy")
+}
+
+// getConfigForClient clones and updates the server's tls config with the
+// appropriate client certificate authorities.
+func getConfigForClient(tlsConfig *tls.Config, ap auth.AccessCache, log logrus.FieldLogger, clusterName string) func(*tls.ClientHelloInfo) (*tls.Config, error) {
+	return func(info *tls.ClientHelloInfo) (*tls.Config, error) {
+		tlsCopy := tlsConfig.Clone()
+
+		pool, err := getCertPool(ap, clusterName)
+		if err != nil {
+			log.WithError(err).Error("Failed to retrieve client CA pool.")
+			return tlsCopy, nil
+		}
+
+		tlsCopy.ClientAuth = tls.RequireAndVerifyClientCert
+		tlsCopy.ClientCAs = pool
+		return tlsCopy, nil
+	}
+}
+
+// getConfigForServer clones and updates the client's tls config with the
+// appropriate server certificate authorities.
+func getConfigForServer(tlsConfig *tls.Config, ap auth.AccessCache, log logrus.FieldLogger, clusterName string) func() (*tls.Config, error) {
+	return func() (*tls.Config, error) {
+		tlsCopy := tlsConfig.Clone()
+
+		pool, err := getCertPool(ap, clusterName)
+		if err != nil {
+			log.WithError(err).Error("Failed to retrieve server CA pool.")
+			return tlsCopy, nil
+		}
+
+		tlsCopy.RootCAs = pool
+		return tlsCopy, nil
+	}
+}
+
+// getCertPool returns a new cert pool from cache if any.
+func getCertPool(ap auth.AccessCache, clusterName string) (*x509.CertPool, error) {
+	pool, _, err := auth.DefaultClientCertPool(ap, clusterName)
```

Comment:
nit: do we actually need to be loading user CAs here?  Might be able to just load host CAs. Ex:

```suggestion
	pool, _, err := auth.ClientCertPool(ap, clusterName, types.HostCA)
```